File: ios/DynamicEntityLayerRef.swift
import ArcGIS
import ExpoModulesCore

/// Operational `DynamicEntityLayer` backed by a real-time data source (a stream service — moving
/// entities that update live). Emits `onConnectionStatusChange` as the data source connects.
public final class DynamicEntityLayerRef: LayerRef {
  let dataSource: DynamicEntityDataSource
  private let pushFeed: PushFeed?
  private var statusTask: Task<Void, Never>?
  private var entityChangeTask: Task<Void, Never>?

  init(props: [String: Any]) {
    let source: DynamicEntityDataSource
    if let custom = props["customSource"] as? [String: Any] {
      let feed = PushFeed()
      let info = DynamicEntityDataSourceInfo(
        entityIDFieldName: custom["entityIdField"] as? String ?? "id",
        fields: buildDynamicEntityFields(custom["fields"] as? [[String: Any]] ?? [])
      )
      self.pushFeed = feed
      source = CustomDynamicEntityDataSource(info: info) { feed }
    } else {
      let urlString = props["streamServiceUrl"] as? String ?? "https://example.invalid"
      self.pushFeed = nil
      source = ArcGISStreamService(url: URL(string: urlString) ?? URL(string: "https://example.invalid")!)
    }
    self.dataSource = source
    super.init(layer: DynamicEntityLayer(dataSource: source))
    observeConnectionStatus()
    observeEntityChanges()
  }

  /// Pushes an observation into a custom data source (no-op for a stream service).
  func pushObservation(_ attributes: [String: Any], _ geometry: [String: Any]) {
    pushFeed?.push(
      geometry: geometryFromDict(geometry),
      attributes: attributes.mapValues(sendableAttribute)
    )
  }

  private func observeConnectionStatus() {
    statusTask = Task { [weak self] in
      guard let self else { return }
      for await status in self.dataSource.$connectionStatus {
        self.emit(event: "onConnectionStatusChange", payload: ["status": connectionStatusString(status)])
      }
    }
  }

  /// Observes `receivedEntities` and `purgedEntities` streams from the data source and emits
  /// `onDynamicEntityChange`. Only entity-lifecycle events are surfaced (`received`/`purged`);
  /// observation-only updates are not forwarded to keep the event rate manageable.
  private func observeEntityChanges() {
    entityChangeTask = Task { [weak self] in
      guard let self else { return }
      // Merge received + purged by spawning two child tasks inside a task group.
      await withTaskGroup(of: Void.self) { group in
        group.addTask { [weak self] in
          guard let self else { return }
          for await info in self.dataSource.receivedEntities {
            guard !Task.isCancelled else { return }
            self.emit(event: "onDynamicEntityChange",
                      payload: dynamicEntityPayload(changeType: "received", entity: info.dynamicEntity))
          }
        }
        group.addTask { [weak self] in
          guard let self else { return }
          for await info in self.dataSource.purgedEntities {
            guard !Task.isCancelled else { return }
            self.emit(event: "onDynamicEntityChange",
                      payload: dynamicEntityPayload(changeType: "purged", entity: info.dynamicEntity))
          }
        }
      }
    }
  }

  /// Returns the data source's currently-tracked dynamic entities (attributes + geometry).
  func queryDynamicEntities() async throws -> [String: Any] {
    let result = try await dataSource.queryDynamicEntities(using: DynamicEntityQueryParameters())
    let entities = Array(result.entities())
    let serialized = entities.map { entity -> [String: Any] in
      var dict: [String: Any] = ["attributes": entity.attributes.mapValues { $0 as Any }]
      if let geometry = entity.geometry { dict["geometry"] = dictFromGeometry(geometry) }
      return dict
    }
    return ["count": entities.count, "entities": serialized]
  }

  override func applyProps(_ changed: [String: Any]) {
    super.applyProps(changed)
    guard let layer = layer as? DynamicEntityLayer else { return }
    if let track = changed["trackDisplay"] as? [String: Any] {
      if let maxObs = track["maximumObservations"] as? NSNumber {
        layer.trackDisplayProperties.maximumObservations = maxObs.intValue
      }
      if let showsPrev = track["showsPreviousObservations"] as? Bool {
        layer.trackDisplayProperties.showsPreviousObservations = showsPrev
      }
    }
    if let filterDict = changed["filter"] as? [String: Any] {
      let filter = ArcGISStreamServiceFilter()
      if let whereClause = filterDict["whereClause"] as? String { filter.whereClause = whereClause }
      if let geometry = (filterDict["geometry"] as? [String: Any]).flatMap(geometryFromDict) { filter.geometry = geometry }
      (dataSource as? ArcGISStreamService)?.filter = filter
    }
    if let purgeDict = changed["purgeOptions"] as? [String: Any] {
      let opts = dataSource.purgeOptions
      if let maxObs = purgeDict["maximumObservations"] as? NSNumber {
        opts.maximumObservations = maxObs.intValue
      }
      if let maxDur = purgeDict["maximumDuration"] as? NSNumber {
        opts.maximumDuration = maxDur.doubleValue
      }
    }
  }

  override public func sharedObjectWillRelease() {
    statusTask?.cancel()
    entityChangeTask?.cancel()
    super.sharedObjectWillRelease()
  }
}

/// Builds a compact payload for the `onDynamicEntityChange` event.
/// Serializes the entity's current attributes snapshot and geometry (if present).
/// Only `received` and `purged` change types are emitted — observation-only attribute
/// updates are skipped; the data source's `receivedEntities` / `purgedEntities` streams
/// fire once per entity lifecycle event, keeping the event rate bounded.
private func dynamicEntityPayload(changeType: String, entity: DynamicEntity) -> [String: Any] {
  var payload: [String: Any] = [
    "changeType": changeType,
    "entityId": entity.id,
    "attributes": entity.attributes.mapValues { $0 as Any },
  ]
  if let geometry = entity.geometry { payload["geometry"] = dictFromGeometry(geometry) }
  return payload
}

/// Maps `ConnectionStatus` to the JS string union.
func connectionStatusString(_ status: ConnectionStatus) -> String {
  switch status {
  case .disconnected: return "disconnected"
  case .connecting: return "connecting"
  case .connected: return "connected"
  case .failed: return "failed"
  @unknown default: return "disconnected"
  }
}

/// A `CustomDynamicEntityFeed` driven imperatively — `push` yields a `newObservation` event.
final class PushFeed: CustomDynamicEntityFeed, @unchecked Sendable {
  typealias Events = AsyncStream<CustomDynamicEntityFeedEvent>
  let events: AsyncStream<CustomDynamicEntityFeedEvent>
  private let continuation: AsyncStream<CustomDynamicEntityFeedEvent>.Continuation

  init() {
    var cont: AsyncStream<CustomDynamicEntityFeedEvent>.Continuation!
    self.events = AsyncStream { cont = $0 }
    self.continuation = cont
  }

  func push(geometry: Geometry?, attributes: [String: any Sendable]) {
    continuation.yield(.newObservation(geometry: geometry, attributes: attributes))
  }
}

/// Converts a JS attribute value to a `Sendable` one (`Sendable` is a marker — it can't be cast at runtime).
private func sendableAttribute(_ value: Any) -> any Sendable {
  if let string = value as? String { return string }
  if let number = value as? NSNumber { return number.doubleValue }
  return String(describing: value)
}

/// Builds `Field`s from JS `{ name, type }` defs for a custom data source (`ArcGIS.Field` — Expo also
/// exposes a `Field` property wrapper, so the namespace is required).
func buildDynamicEntityFields(_ defs: [[String: Any]]) -> [ArcGIS.Field] {
  defs.map { def in
    ArcGIS.Field(type: dynamicEntityFieldType(def["type"] as? String), name: def["name"] as? String ?? "", alias: "")
  }
}

private func dynamicEntityFieldType(_ type: String?) -> FieldType {
  switch type {
  case "int32": return .int32
  case "int64": return .int64
  case "float64": return .float64
  case "date": return .date
  default: return .text
  }
}

File: android/src/main/java/expo/modules/arcgis/DynamicEntityLayerRef.kt
package expo.modules.arcgis

import com.arcgismaps.data.Field
import com.arcgismaps.data.FieldType
import com.arcgismaps.mapping.layers.DynamicEntityLayer
import com.arcgismaps.mapping.layers.Layer
import com.arcgismaps.realtime.ArcGISStreamService
import com.arcgismaps.realtime.ArcGISStreamServiceFilter
import com.arcgismaps.realtime.ConnectionStatus
import com.arcgismaps.realtime.CustomDynamicEntityDataSource
import com.arcgismaps.realtime.DynamicEntityDataSource
import com.arcgismaps.realtime.DynamicEntityDataSourceInfo
import com.arcgismaps.realtime.DynamicEntityQueryParameters
import expo.modules.kotlin.AppContext
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.SharedFlow
import kotlinx.coroutines.flow.asSharedFlow
import kotlinx.coroutines.launch

/**
 * Operational [DynamicEntityLayer] backed by a real-time data source (a stream service — moving
 * entities that update live). Emits `onConnectionStatusChange` as the data source connects.
 */
class DynamicEntityLayerRef(appContext: AppContext, props: Map<String, Any?>) : LayerRef(appContext) {
  // Non-null only in custom-source mode — observations pushed from JS are emitted into this flow.
  private val pushFlow: MutableSharedFlow<CustomDynamicEntityDataSource.FeedEvent>? =
    if (props["customSource"] != null) MutableSharedFlow(extraBufferCapacity = 64) else null
  val dataSource: DynamicEntityDataSource = buildDataSource(props, pushFlow)
  override val layer: Layer = DynamicEntityLayer(dataSource)
  private val scope = CoroutineScope(Dispatchers.Main + SupervisorJob())

  init {
    scope.launch {
      dataSource.connectionStatus.collect { status ->
        emit("onConnectionStatusChange", mapOf("status" to connectionStatusString(status)))
      }
    }
    // Emit entity-received events (new/updated observation arrived for an entity).
    scope.launch {
      dataSource.dynamicEntityReceivedEvent.collect { info ->
        emit("onDynamicEntityChange", dynamicEntityPayload("received", info.dynamicEntity))
      }
    }
    // Emit entity-purged events (entity evicted by purge rules).
    scope.launch {
      dataSource.dynamicEntityPurgedEvent.collect { info ->
        emit("onDynamicEntityChange", dynamicEntityPayload("purged", info.dynamicEntity))
      }
    }
  }

  /** Pushes an observation into a custom data source (no-op for a stream service). */
  fun pushObservation(attributes: Map<String, Any?>, geometry: Map<String, Any?>) {
    val flow = pushFlow ?: return
    val geom = geometryFromDict(geometry) ?: return
    flow.tryEmit(CustomDynamicEntityDataSource.FeedEvent.NewObservation(geom, attributes))
  }

  /**
   * Returns the observation history for the entity with the given track id, newest first,
   * capped at [max] entries (default 100). Each observation carries its own `attributes` and
   * `geometry` snapshot at the time it was received.
   */
  suspend fun queryObservations(entityId: String, max: Int = 100): List<Map<String, Any?>> {
    val result = dataSource.queryDynamicEntities(listOf(entityId)).getOrThrow()
    val entity = result.toList().firstOrNull()
      ?: return emptyList()
    return entity.getObservations(max).map { obs ->
      val entry = mutableMapOf<String, Any?>(
        "attributes" to obs.attributes.toMap(),
      )
      obs.geometry?.let { entry["geometry"] = dictFromGeometry(it) }
      entry
    }
  }

  /** Returns the data source's currently-tracked dynamic entities (attributes + geometry). */
  suspend fun queryDynamicEntities(): Map<String, Any?> {
    val result = dataSource.queryDynamicEntities(DynamicEntityQueryParameters()).getOrThrow()
    val entities = result.toList()
    return mapOf(
      "count" to entities.size,
      "entities" to entities.map { entity ->
        mapOf(
          "attributes" to entity.attributes,
          "geometry" to entity.geometry?.let { dictFromGeometry(it) },
        )
      },
    )
  }

  override fun applyProps(changed: Map<String, Any?>) {
    applyCommonProps(changed)
    val entityLayer = layer as? DynamicEntityLayer ?: return
    (changed["trackDisplay"] as? Map<*, *>)?.let { track ->
      (track["maximumObservations"] as? Number)?.toInt()?.let {
        entityLayer.trackDisplayProperties.maximumObservations = it
      }
      (track["showsPreviousObservations"] as? Boolean)?.let {
        entityLayer.trackDisplayProperties.showPreviousObservations = it
      }
    }
    (changed["filter"] as? Map<*, *>)?.let { filterDict ->
      val filter = ArcGISStreamServiceFilter().apply {
        (filterDict["whereClause"] as? String)?.let { whereClause = it }
        (filterDict["geometry"] as? Map<*, *>)?.let { g -> geometryFromDict(g)?.let { geometry = it } }
      }
      (dataSource as? ArcGISStreamService)?.filter = filter
    }
    (changed["purgeOptions"] as? Map<*, *>)?.let { purgeDict ->
      val opts = dataSource.purgeOptions
      (purgeDict["maximumObservations"] as? Number)?.let { opts.maximumObservations = it.toLong() }
      (purgeDict["maximumDuration"] as? Number)?.let { opts.maximumDuration = it.toDouble() }
    }
  }

  override fun deallocate() {
    scope.cancel()
    super.deallocate()
  }
}

/** Maps [ConnectionStatus] to the JS string union. */
fun connectionStatusString(status: ConnectionStatus): String = when (status) {
  is ConnectionStatus.Disconnected -> "disconnected"
  is ConnectionStatus.Connecting -> "connecting"
  is ConnectionStatus.Connected -> "connected"
  is ConnectionStatus.Failed -> "failed"
}

/**
 * Builds a compact payload for the `onDynamicEntityChange` event.
 * Only `received` and `purged` change types are emitted (observation-only updates are
 * captured within `dynamicEntityReceivedEvent` which fires per-entity arrival, not per
 * observation, so the event rate is bounded to entity lifecycle changes).
 * Geometry is serialized only when present; attributes are passed as-is (the map returned
 * by the SDK is a shallow snapshot of the current entity attributes).
 */
private fun dynamicEntityPayload(changeType: String, entity: com.arcgismaps.realtime.DynamicEntity): Map<String, Any?> {
  val payload = mutableMapOf<String, Any?>(
    "changeType" to changeType,
    "entityId" to entity.id,
    "attributes" to entity.attributes.toMap(),
  )
  entity.geometry?.let { payload["geometry"] = dictFromGeometry(it) }
  return payload
}

/** Builds the real-time data source: a custom feed (push from JS) or a stream service. */
private fun buildDataSource(
  props: Map<String, Any?>,
  pushFlow: MutableSharedFlow<CustomDynamicEntityDataSource.FeedEvent>?,
): DynamicEntityDataSource {
  val custom = props["customSource"] as? Map<*, *>
  if (custom != null && pushFlow != null) {
    val info = DynamicEntityDataSourceInfo(
      custom["entityIdField"] as? String ?: "id",
      buildDynamicEntityFields(custom["fields"] as? List<*>),
    )
    return CustomDynamicEntityDataSource(object : CustomDynamicEntityDataSource.EntityFeedProvider {
      override val feed: SharedFlow<CustomDynamicEntityDataSource.FeedEvent> = pushFlow.asSharedFlow()
      override suspend fun onLoad(): DynamicEntityDataSourceInfo = info
      override suspend fun onConnect() {}
      override suspend fun onDisconnect() {}
    })
  }
  return ArcGISStreamService(props["streamServiceUrl"] as? String ?: "")
}

/** Builds [Field]s from JS `{ name, type }` defs for a custom data source. */
private fun buildDynamicEntityFields(defs: List<*>?): List<Field> =
  (defs ?: emptyList<Any?>()).filterIsInstance<Map<*, *>>().map { def ->
    Field(dynamicEntityFieldType(def["type"] as? String), def["name"] as? String ?: "", "", 0, null, true, true)
  }

private fun dynamicEntityFieldType(type: String?): FieldType = when (type) {
  "int32" -> FieldType.Int32
  "int64" -> FieldType.Int64
  "float64" -> FieldType.Float64
  "date" -> FieldType.Date
  else -> FieldType.Text
}
